### ISSUE TITLE
Tensorflow 

### ISSUE THREAD

**Issue by naatje80** [tensorflow/tensorflow#25312]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Scientific Linux 7.6, but also tested on Centos 7.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary): source
- TensorFlow version: 1.12.0
- Python version: 2.6
- Installed using virtualenv? pip? conda?: manually compiling using bazel
- Bazel version (if compiling from source): 1.16.1, but also tried 1.15.0 and 1.18.0
- GCC/Compiler version (if compiling from source): 4.8.5
- CUDA/cuDNN version: Cuda 9 CuDNN 7
- GPU model and memory: Compute Cluster: multpile Tesla K80 (24GB) or Tesla P100 (16GB).

**Describe the problem**
During compilation I'm encountering the following ERROR messages:
[205 / 208] Compiling tensorflow/contrib/image/kernels/adjust_hsv_in_yiq_op_gpu.cu.cc; 10s local
ERROR: /tmp/Tensorflow/PACKAGES/tensorflow/tensorflow/contrib/image/BUILD:115:1: undeclared inclusion(s) in rule '//tensorflow/contrib/image:python/ops/_distort_image_ops_gpu':
this rule is missing dependency declarations for the following files included by 'tensorflow/contrib/image/kernels/adjust_hsv_in_yiq_op_gpu.cu.cc':
  '/tmp/bazel/userid_output/external/eigen_archive/Eigen/Core'
  '/tmp/bazel/userid_output/external/eigen_archive/Eigen/src/Core/util/DisableStupidWarnings.h'
  '/tmp/bazel/userid_output/external/eigen_archive/Eigen/src/Core/util/ReenableStupidWarnings.h
However, when I add ---verbose_failures to bazel and execute the failing compilation job manually to debug the issue, the compilation is executed correctly. So I'm assuming only some check of bazel is failing, although the compilation itself could be executed correctly.

Restarting the build (after my manually compilation) will also continue, until the next job where Eigen is used again. The build is failing with a similar error. Previous versions, including 1.11.0, are not encountering this issue. I could not find any noticeable changes in the Eigen build files for bazel or related to Eigen. Therefore I'm currently out of ideas.....

**Provide the exact sequence of commands / steps that you executed before running into the problem**
bazel --output_base=/tmp/bazel/userid_output build --jobs 12 -c opt --copt=-mavx2 --copt=-O --copt=-msse4.2 --copt=-mfma -config=cuda //tensorflow/tools/pip_package:build_pip_package
(I'm using bazel output_base because my home directory is on NFS)

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gunan** [comment]
This is really interesting. I have never seen this issue.
While this may cost you quite some time, could you try running:
`bazel clean --expunge`
Then rerunning configure, then build?

Also, did you modify the source code at all?

**Comment by Jojuh** [comment]
I'm having the same issue, CentOS 7, Tensorflow 1.12, GCC 7.3.0, CUDA 9, cuDNN 7. Something breaks eigen include paths when using --output_base. I found this when trying to build using EasyBuild and replicated the steps when building manually to chase what was failing the build. The following works for me:

`bazel build --compilation_mode=opt --action_env=PYTHONPATH --config=opt --config=cuda --distinct_host_configuration=false --config=mkl //tensorflow/tools/pip_package:build_pip_package`

Adding --output_base=/home/jojuh/test gives me (the actual rule name varies as these are build parallel?):

```
ERROR: /home/jojuh/tensorflow-1.12.0/tensorflow/tools/pip_package/BUILD:48:1 undeclared inclusion(s) in rule '//tensorflow/core:lib_internal_impl':
this rule is missing dependency declarations for the following files included by 'tensorflow/core/lib/random/distribution_sampler.cc':
  '/home/jojuh/test/external/eigen_archive/Eigen/Core'
  '/home/jojuh/test/external/eigen_archive/Eigen/src/Core/util/DisableStupidWarnings.h'
  '/home/jojuh/test/external/eigen_archive/Eigen/src/Core/util/Macros.h'
  '/home/jojuh/test/external/eigen_archive/Eigen/src/Core/util/MKL_support.h'
  ....
```

Apparently eigen is included via a wrapper (//third_party/eigen3) which points to eigen_archive somehow (https://github.com/tensorflow/tensorflow/issues/3743#issuecomment-239580919)

I'm new to Bazel and this is currently blocking our progress so I'm happy to provide more information if necessary.

**Comment by naatje80** [comment]
Thank you both for your reply/additional information. 
@gunan My buildscript already executes the bazel clean --expunge step to ensure it starts from scratch. The only source change I'm making is adding the --output_base=${BAZEL_OUTPUT_BASE} option to the bazel command in configure.py:
**sed -i -r "s#\['bazel', '--batch', '--bazelrc=/dev/null', 'version'\]#\['bazel'    ,  '--output_base=${BAZEL_OUTPUT_BASE}', '--batch', '--bazelrc=/dev/null', 'version'\]#    " configure.py**

@Jojuh The actual rules indeed very do to the parallel build. You could force a build using a single thread by adding --jobs 1to bazel for debugging. But it takes a while before the actual error pops up.

As explained before, I also require the --output_base because the shares/disks I have write access to are located on external file servers (except the /tmp folder, which I now use for the bazel output base).

**Comment by Jojuh** [comment]
I finally got my build to complete, had to add compiler flags for the crosstool_wrapper_driver_is_not_gcc. Not absolutely sure if I understood correctly, but by passing -MMD instead of -MD to the compiler makes it ignore system headers -> no undeclared dependency errors. Would be happy if someone would ELI5 this to me.

```
diff -ru tensorflow-1.12.0.orig/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl tensorflow-1.12.0/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl
--- tensorflow-1.12.0.orig/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl      2018-11-02 03:35:10.000000000 +0200
+++ tensorflow-1.12.0/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl  2019-02-18 09:20:02.114287752 +0200
@@ -258,6 +261,7 @@
   cpu_compiler_flags = [flag for flag in sys.argv[1:]
                              if not flag.startswith(('--cuda_log'))]

+  cpu_compiler_flags = ['-MMD' if flag == '-MD' else flag for flag in cpu_compiler_flags]
   return subprocess.call([CPU_COMPILER] + cpu_compiler_flags)

 if __name__ == '__main__':
```

This led me to the solution: https://stackoverflow.com/questions/35256110/tensorflow-build-fails-with-missing-dependency-error

**Comment by naatje80** [comment]
@Jojuh Thank you, I will try your suggestion. What I do not understand is how this could be an issue for bazel and not for the compilation step itself. Because, like mentioned before, if I executed the failing step manually, it compiles without any issue. But hopefully the additional information will help to find the issue, because it was not an issue before release 1.12.0.....

**Comment by tlalexander** [comment]
I tried @Jojuh 's solution but it did not solve the problem for me. What did work in my case was to make a symbolic link from the normal cache directory to the folder on another disk I was trying to use for the output directory. Then I could run without the --output_base flag but still use the storage location I wanted. I was also running out of memory, which I fixed by allocating some swap. Not sure if that was related.

For the workaround:
cd ~/
mv .cache/bazel /path/to/new/folder/
ln -s /path/to/new/folder/bazel .cache/bazel

It seems like there is a clear bug with the --output_base flag, but since the above fix (hopefully I applied it properly) didn't work, I wanted to share my workaround.

Details:
Tensorflow 1.12
Cuda 9.0
CPU aarch64 (Jetson TX1)
Python 3.5.2
Ubuntu from JetPack 3.3

Where I was also plagued with "undeclared inclusion(s) in rule '//tensorflow/core:lib_internal_impl':
this rule is missing dependency declarations for the following files included by ...." errors followed by a list of files from "...external/eigen_archive/Eigen/..." while using the --output_base flag.

In my case I was also running out of memory, though I don't know if that triggered this issue. I fixed the memory issues by allocating some swap space.

After the workaround I mentioned and adding the swap memory, I've been able to sustain build for over an hour (it's an arm system). Previously the errors would come up after only ten minutes, so this seems like progress...

**Comment by naatje80** [comment]
The patch supplied by @Jojuh did indeed also not fix the issue for my. I will now also try the suggestion mentioned by @tlalexander. But clearly something must hast changed, because the --output_base option was never an issue before.

**Comment by naatje80** [comment]
@Jojuh: I was indeed also able to compile the module by creating a symbolic link to temp instead of using the output_base. @gunan hopefully this is helpful in locating the actual issue. But at least we have a workaround for now.

**Comment by naatje80** [comment]
This issue seems to be fixed for version 1.13.1. I was able to use --output_base again.

**Comment by placidchat** [comment]
Could someone explain the -MMD -MD difference again?

**Comment by gunan** [comment]
Closing as it is reported that the issue is fixed in the newer versions.
